### ISSUE TITLE
🧪 [Experiment] TokenView

### ISSUE THREAD

**Issue by niels9001** [CommunityToolkit/Labs-Windows#427]
### Approved from Discussion

https://github.com/CommunityToolkit/Labs-Windows/discussions/323

### Problem Statement

An easy way to visualize filters / selected data / options.

### Overview

This experiment adds the following components:
 - **TokenView**
 - **TokenItem**

### Using

You can try it out via the NuGet Packages here:
 - UWP: https://dev.azure.com/dotnet/CommunityToolkit/_artifacts/feed/CommunityToolkit-Labs/NuGet/CommunityToolkit.Labs.Uwp.TokenView
 - WinUI 3: https://dev.azure.com/dotnet/CommunityToolkit/_artifacts/feed/CommunityToolkit-Labs/NuGet/CommunityToolkit.Labs.WinUI.TokenView
Read more about [Preview Packages here](https://aka.ms/wct/wiki/previewpackages).

**Code**

**Result**

**Documentation & samples**

### Implementation Requirements

- [X] Working Prototype
- [X] Feature Complete
- [X] Documentation
- [X] Samples
- [ ] Unit Tests
- [ ] Community Feedback / Usage Testimonies

### Tested Platforms

- [ ] UWP
- [ ] WinAppSDK / WinUI 3
- [ ] Web Assembly (WASM)
- [ ] Android
- [ ] iOS
- [ ] MacOS
- [ ] Linux / GTK

### Technical Review

- [ ] Accessibility Audit
- [ ] API/Naming Review
- [ ] Code Quality/Style
- [ ] Dependency Review
- [ ] Design/Style Review
- [ ] Final Approval

### Community Help?

None

**Comment by jp-weber** [comment]
I have noticed a few small things:

The corner radius could not be overwritten uniformly and was defined differently than in the standard.
Static corner radius (99):
https://github.com/CommunityToolkit/Labs-Windows/blob/main/components/TokenView/src/TokenItem/TokenItem.xaml#L294

Definition of TokenItemCornerRadius
https://github.com/CommunityToolkit/Labs-Windows/blob/main/components/TokenView/src/TokenItem/TokenItem.xaml#L129
```xaml
    <x:Double x:Key="TokenItemCornerRadius">16</x:Double>
```
Standard e.g. https://github.com/microsoft/microsoft-ui-xaml/blob/4041a77fa1468062ef45bfe7a136aeca6e9044f2/dev/ProgressBar/ProgressBar_themeresources.xaml#L40
```xaml
<CornerRadius x:Key="ProgressBarCornerRadius">1.5</CornerRadius>
```

The foreground of the remove Button is always black:
![grafik](https://user-images.githubusercontent.com/10946850/233830796-8e9c0e15-2407-4503-933f-6f2b66551a8f.png)

**Comment by niels9001** [comment]
> I have noticed a few small things:
> 
> The corner radius could not be overwritten uniformly and was defined differently than in the standard. Static corner radius (99): https://github.com/CommunityToolkit/Labs-Windows/blob/main/components/TokenView/src/TokenItem/TokenItem.xaml#L294
> 
> Definition of TokenItemCornerRadius https://github.com/CommunityToolkit/Labs-Windows/blob/main/components/TokenView/src/TokenItem/TokenItem.xaml#L129
> 
> ```
>     <x:Double x:Key="TokenItemCornerRadius">16</x:Double>
> ```
> 
> Standard e.g. https://github.com/microsoft/microsoft-ui-xaml/blob/4041a77fa1468062ef45bfe7a136aeca6e9044f2/dev/ProgressBar/ProgressBar_themeresources.xaml#L40
> 
> ```
> <CornerRadius x:Key="ProgressBarCornerRadius">1.5</CornerRadius>
> ```
> 
> The foreground of the remove Button is always black: ![grafik](https://user-images.githubusercontent.com/10946850/233830796-8e9c0e15-2407-4503-933f-6f2b66551a8f.png)

@jp-weber Great feedback! I'll have a look at this!

**Comment by yaira2** [comment]
Is it possible to set an item source for this control?

**Comment by niels9001** [comment]
> Is it possible to set an item source for this control?

Yep.. see a sample here: https://github.com/CommunityToolkit/Labs-Windows/blob/main/components/TokenView/samples/TokenViewItemsSourceSample.xaml

**Comment by Jay-o-Way** [comment]
@niels9001 a few weeks ago I sent you an email with a question about coding. Did you see that? Would love to hear the answer. You'll understand what I mean when you see it.

**Comment by Jay-o-Way** [comment]
Question: why is radius = 16? Generally, controls would use 8. It looks nice and semi-circle with this size, but what if the OS uses a bigger _font_ size? (Not to be confused with general scaling)

**Comment by jp-weber** [comment]
I am currently working on the exemplary implementation in my application. But if I use a TokenItem in the DataTemplate , it looks like a TokenItem is generated in a TokenItem. Have I missed something?
![grafik](https://github.com/CommunityToolkit/Labs-Windows/assets/10946850/d0810ee4-7c18-4a48-92a2-20a85a865922)

```xaml
<labs:TokenView 
    ItemsSource="{x:Bind ViewModel.TempAccountCategoryList,Mode=OneWay}"
    IsItemClickEnabled="True"
    SelectionMode="Multiple" 
    ItemClick="TokenView_ItemClick" >
    <labs:TokenView.ItemTemplate>
        <DataTemplate x:DataType="models:CategoryModel">
            <labs:TokenItem 
                Content="{x:Bind Name,Mode=OneWay}" 
                IsSelected="{x:Bind IsSelected,Mode=TwoWay}">
                <labs:TokenItem.Icon>
                    <FontIcon Glyph="{x:Bind Glyph,Mode=OneWay}"/>
                </labs:TokenItem.Icon>
            </labs:TokenItem>
        </DataTemplate>
    </labs:TokenView.ItemTemplate>
</labs:TokenView>
```

**Comment by niels9001** [comment]
> I am currently working on the exemplary implementation in my application. But if I use a TokenItem in the DataTemplate , it looks like a TokenItem is generated in a TokenItem. Have I missed something? ![grafik](https://user-images.githubusercontent.com/10946850/239701729-d0810ee4-7c18-4a48-92a2-20a85a865922.png)
> 
> ```
> <labs:TokenView 
>     ItemsSource="{x:Bind ViewModel.TempAccountCategoryList,Mode=OneWay}"
>     IsItemClickEnabled="True"
>     SelectionMode="Multiple" 
>     ItemClick="TokenView_ItemClick" >
>     <labs:TokenView.ItemTemplate>
>         <DataTemplate x:DataType="models:CategoryModel">
>             <labs:TokenItem 
>                 Content="{x:Bind Name,Mode=OneWay}" 
>                 IsSelected="{x:Bind IsSelected,Mode=TwoWay}">
>                 <labs:TokenItem.Icon>
>                     <FontIcon Glyph="{x:Bind Glyph,Mode=OneWay}"/>
>                 </labs:TokenItem.Icon>
>             </labs:TokenItem>
>         </DataTemplate>
>     </labs:TokenView.ItemTemplate>
> </labs:TokenView>
> ```

The `ItemTemplate` is part of the `TokenItem` already, so no need to add a `TokenItem` within the `DataTemplate`. So you should be able to use e.g. a `Grid` and add a `TextBlock` to bind the `Name`. See: https://github.com/CommunityToolkit/Labs-Windows/blob/255e9a5bd1c3e68af4ec2dbfae8e4d00dc5e1d37/components/TokenView/samples/TokenViewItemsSourceSample.xaml#L19

`TokenView` derives from `ListViewBase`, so you should be able to re-use `ListView` specific samples as well (there are a lot of `ListView` binding examples out there),

**Comment by jp-weber** [comment]
> `TokenView` derives from `ListViewBase`, so you should be able to re-use `ListView` specific samples as well (there are a lot of `ListView` binding examples out there),

Thanks you are right. I hadn't thought about that at all and was focused on the TokenItem because I wanted to bind 'IsSelected', which is of course feasible with the ListViewBase with SelectedItems.

**Comment by vilgotf** [comment]
Removing items does not work with AoT 😢: https://github.com/CommunityToolkit/Labs-Windows/blob/b71dbf1c72dcef89a6f94b9c4cf57b2ab0ec0e1d/components/TokenView/src/TokenView/TokenView.ItemsSource.cs#L20-L29

**Comment by weitzhandler** [comment]
[Gallery: TokenView page freezes app](https://github.com/CommunityToolkit/Windows/issues/596)

**Comment by weitzhandler** [comment]
Can we have an 'Add' button as part of the control?

**Comment by PhoenixNil** [comment]
<img width="65" height="60" alt="Image" src="https://github.com/user-attachments/assets/528ee4b3-5c15-4739-8616-6dd985458000" /> The edges are very jagged and muddy color if u can zoom in . for this tokenview so ugly Implemented I don't understand why everyone ignore this problem

**Comment by weitzhandler** [comment]
Can you please do something so that we can build the Labs projecs from CI? Currently it's blocked by GitHub:

```
# Direct test shows DNS refuses blob storage domains:
$ nslookup ixuvsblobprodwus2132.blob.core.windows.net
** server can't find ixuvsblobprodwus2132.blob.core.windows.net: REFUSED

# Even general blob storage domain is blocked:
$ ping blob.core.windows.net
ping: blob.core.windows.net: No address associated with hostname
```
